### DISCUSSION TITLE
Need help running server on Android with OpenCL support

### DISCUSSION THREAD

**Discussion by DGdev91** [ggml-org/llama.cpp#5507] [Q&A]
As written on the readme, for running llama.cpp with CLBlast support on Android we need to add this before launching: export LD_LIBRARY_PATH=/vendor/lib64:$LD_LIBRARY_PATH

This works fine for main, but running server does not work, because it cannot find some required libs for networking.

That's the error:
CANNOT LINK EXECUTABLE "./server": cannot locate symbol "__emutls_get_address" referenced by "/data/data/com.termux/files/home/llama.cpp/server"...

Someone has some idea on how to make that work?

**Comment by restartxx** [comment]
Same here...

**Reply by punithrudrappa** [reply]
I'm having the same issue. I think we need to raise an issue for this.

**Reply by DGdev91** [reply]
Actually, i did. But it got almost immediatly closed.
https://github.com/termux/termux-packages/issues/19292

Termux has his own libs wich differ from android's, and using LD_LIBRARY_PATH isn't supported. Can't really argue with that.

The suggestion i got as answer didn't work for me, but i'm not 100% sure i did that correctly, that's why i still didn't replied

**Comment by Henri-J-Norden** [comment]
I think only libOpenCL.so is needed, so you can copy that to another dir and use that dir as your LD_LIBRARY_PATH - avoiding overriding any other libraries

E.g just copy it to current dir (~/llama.cpp/build/bin):
`cp /vendor/lib64/libOpenCL.so .`
Then you can either export the var like before (use absolute path for that), or just prefix it to the command:
`LD_LIBRARY_PATH=. clinfo -l`
↑ This now reports correctly for me:
`Platform #0: QUALCOMM Snapdragon(TM)                                               -- Device #0: QUALCOMM Adreno(TM)`
So, to run llama.cpp:
`LD_LIBRARY_PATH=. ./server -m model.gguf`

Though I'm not sure if this really worked (or if I went wrong somewhere else), because tokens/sec performance does not seem better than the version compiled without OpenCL, but I need to do more testing... maybe it works better for you?

**Reply by punithrudrappa** [reply]
"/data/data/com.termux/files/home/llama.cpp/libOpenCL.so" has bad ELF magic: 00000000

I don't understand how do I go around this.

**Reply by DGdev91** [reply]
Well, i didn't even tought that could actually work. But it does, apparently.
Seems like libOpenCL.so is indeed enough.

The reason you aren't noticing any difference is most likely because you need to use the --n-gpu-layers (or just -ngl) parameter.

My start script looks like this:

```
export LD_LIBRARY_PATH=.:$LD_LIBRARY_PATH
./server -t 8 -ngl 200 -m path_to_model.gguf
```

**Reply by DGdev91** [reply]
> "/data/data/com.termux/files/home/llama.cpp/libOpenCL.so" has bad ELF magic: 00000000
> 
> I don't understand how do I go around this.

i guess it depends on the phone. in the guide it's mentioned that some phones need to use /system/vendor/lib64 instead of /vendor/lib64. you can try.

Also, are you using the fdroid build of fdroid or the one from the play store? if it's the second, you need to use the fdroid one.

**Reply by Danmoreng** [reply]
I tried everything of the above, but my clinfo does not detect any platforms. Running on a Samsung S24 with Exynos 2400 chip. Anyone probably can help me?

**Reply by woachk** [reply]
For the Galaxy phones with Xclipse GPUs you have to do: 

`LD_LIBRARY_PATH=/vendor/lib64:/vendor/lib64/hw` as Samsung screwed up the relpaths for their OpenCL impl
